// @blueprintjs/docs-theme/lib/esnext/components/baseExample.js
/*
 * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Utils } from "@blueprintjs/core";
import classNames from "classnames";
import * as React from "react";
export const WARNING_BASE_EXAMPLE_DEPRECATED = `[Blueprint] BaseExample is deprecated and will be removed in the next beta. Compose new Example component instead of extending BaseExample.`;
/**
 * Starter class for all React example components.
 * Examples and options are rendered into separate containers.
 * @deprecated
 */
export class BaseExample extends React.Component {
    constructor() {
        super(...arguments);
        // Can't put this in state, because the state typing is generic.
        this.hasDelayedBeforeInitialRender = false;
        this.hasCompletedInitialRender = false;
    }
    shouldComponentUpdate(nextProps, nextState) {
        return (
        // HACKHACK: Permit one redundant re-render after the inital delay. (This will be the first proper render of
        // the component.)
        (this.hasDelayedBeforeInitialRender && !this.hasCompletedInitialRender) ||
            // Now, mimic PureComponent shouldComponentUpdate behavior:
            !Utils.shallowCompareKeys(this.props, nextProps) ||
            !Utils.shallowCompareKeys(this.state, nextState));
    }
    componentWillMount() {
        // HACKHACK: The docs app suffers from a Flash of Unstyled Content that causes some 'width: 100%' examples to
        // render incorrectly, because they mis-measure the horizontal space available to them. Until that bug is squashed,
        // this is the workaround: delay initial render with a requestAnimationFrame.
        requestAnimationFrame(() => {
            this.hasDelayedBeforeInitialRender = true;
            this.forceUpdate();
        });
    }
    componentDidMount() {
        console.warn(WARNING_BASE_EXAMPLE_DEPRECATED);
    }
    componentDidUpdate(_nextProps, _nextState) {
        // HACKHACK: Initial render happens as an *update* due to our requestAnimationFrame shenanigans, not as a mount.
        // Once we've rendered initially, set this flag so that shouldComponentUpdate logic will return to its normal
        // PureComponent-style logic, ignoring these flags henceforth.
        if (!this.hasCompletedInitialRender) {
            this.hasCompletedInitialRender = true;
        }
    }
    render() {
        // HACKHACK: This is the other required piece. Don't let any React nodes into the DOM until the
        // requestAnimationFrame delay has elapsed. This prevents shouldComponentUpdate snafus at lower levels.
        if (!this.hasDelayedBeforeInitialRender) {
            return null;
        }
        return (React.createElement("div", { className: classNames("docs-example-frame", this.className), "data-example-id": this.props.id },
            React.createElement("div", { className: "docs-example" }, this.renderExample()),
            React.createElement("div", { className: "docs-example-options" }, this.actuallyRenderOptions())));
    }
    /**
     * Render the example element. Return any valid React node.
     */
    renderExample() {
        return undefined;
    }
    /**
     * Render the options controls. Return a single element for simple mode, or an array of arrays
     * of elements to generate columns: each array will be its own column. When using array mode,
     * the inner elements will each need the `key` prop.
     */
    renderOptions() {
        return [];
    }
    actuallyRenderOptions() {
        const options = this.renderOptions();
        if (Array.isArray(options)) {
            return options.map((column, i) => (React.createElement("div", { className: "docs-react-options-column", key: i }, column)));
        }
        else {
            return options;
        }
    }
}
/** Event handler that exposes the target element's value as a boolean. */
export function handleBooleanChange(handler) {
    return (event) => handler(event.target.checked);
}
/** Event handler that exposes the target element's value as a string. */
export function handleStringChange(handler) {
    return (event) => handler(event.target.value);
}
/** Event handler that exposes the target element's value as a number. */
export function handleNumberChange(handler) {
    return handleStringChange(value => handler(+value));
}
