// @blueprintjs/docs-theme/lib/esnext/components/navigator.js
/*
 * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Classes, Icon, MenuItem, Utils } from "@blueprintjs/core";
import { Omnibar } from "@blueprintjs/select";
import { filter } from "fuzzaldrin-plus";
import * as React from "react";
import { eachLayoutNode } from "../common/utils";
const NavOmnibar = Omnibar.ofType();
const INPUT_PROPS = { placeholder: "Fuzzy search headings..." };
export class Navigator extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.filterMatches = (query, items) => filter(items, query, {
            key: "route",
            maxInners: items.length / 5,
            maxResults: 10,
            pathSeparator: "/",
            usePathScoring: true,
        });
        this.renderItem = (section, props) => {
            if (!props.modifiers.matchesPredicate) {
                return null;
            }
            // insert caret-right between each path element
            const pathElements = section.path.reduce((elems, el) => {
                elems.push(el, React.createElement(Icon, { key: el, icon: "caret-right" }));
                return elems;
            }, []);
            pathElements.pop();
            const text = (React.createElement(React.Fragment, null,
                React.createElement("div", null, section.title),
                React.createElement("small", { className: Classes.TEXT_MUTED }, pathElements)));
            return (React.createElement(MenuItem, { active: props.modifiers.active, href: "#" + section.route, key: section.route, multiline: true, onClick: props.handleClick, text: text }));
        };
        // updating location.hash will trigger hashchange event, which Documentation will receive and use to navigate.
        this.handleItemSelect = (item) => {
            location.hash = item.route;
            this.props.onClose();
        };
    }
    componentDidMount() {
        this.sections = [];
        eachLayoutNode(this.props.items, (node, parents) => {
            if (Utils.safeInvoke(this.props.itemExclude, node) === true) {
                // ignore excluded item
                return;
            }
            const { route, title } = node;
            const path = parents.map(p => p.title).reverse();
            this.sections.push({ path, route, title });
        });
    }
    render() {
        if (!this.sections) {
            return null;
        }
        return (React.createElement(NavOmnibar, { className: "docs-navigator-menu", inputProps: INPUT_PROPS, itemListPredicate: this.filterMatches, isOpen: this.props.isOpen, items: this.sections, itemRenderer: this.renderItem, onItemSelect: this.handleItemSelect, onClose: this.props.onClose, resetOnSelect: true }));
    }
}
