// @blueprintjs/table/lib/esnext/headers/columnHeaderCell.js
/*
 * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as tslib_1 from "tslib";
import classNames from "classnames";
import * as React from "react";
import { polyfill } from "react-lifecycles-compat";
import { AbstractPureComponent2, Icon, Popover, Position, Utils as CoreUtils, } from "@blueprintjs/core";
import * as Classes from "../common/classes";
import { columnInteractionBarContextTypes } from "../common/context";
import { LoadableContent } from "../common/loadableContent";
import { HeaderCell } from "./headerCell";
export function HorizontalCellDivider() {
    return React.createElement("div", { className: Classes.TABLE_HORIZONTAL_CELL_DIVIDER });
}
let ColumnHeaderCell = class ColumnHeaderCell extends AbstractPureComponent2 {
    constructor() {
        super(...arguments);
        this.state = {
            isActive: false,
        };
        this.handlePopoverOpened = () => this.setState({ isActive: true });
        this.handlePopoverClosing = () => this.setState({ isActive: false });
    }
    /**
     * This method determines if a `MouseEvent` was triggered on a target that
     * should be used as the header click/drag target. This enables users of
     * this component to render fully interactive components in their header
     * cells without worry of selection or resize operations from capturing
     * their mouse events.
     */
    static isHeaderMouseTarget(target) {
        return (target.classList.contains(Classes.TABLE_HEADER) ||
            target.classList.contains(Classes.TABLE_COLUMN_NAME) ||
            target.classList.contains(Classes.TABLE_INTERACTION_BAR) ||
            target.classList.contains(Classes.TABLE_HEADER_CONTENT));
    }
    render() {
        const { 
        // from IColumnHeaderCellProps
        enableColumnReordering, isColumnSelected, menuIcon, 
        // from IColumnNameProps
        name, nameRenderer, 
        // from IHeaderProps
        ...spreadableProps } = this.props;
        const classes = classNames(spreadableProps.className, Classes.TABLE_COLUMN_HEADER_CELL, {
            [Classes.TABLE_HAS_INTERACTION_BAR]: this.context.enableColumnInteractionBar,
            [Classes.TABLE_HAS_REORDER_HANDLE]: this.props.reorderHandle != null,
        });
        return (React.createElement(HeaderCell, Object.assign({ isReorderable: this.props.enableColumnReordering, isSelected: this.props.isColumnSelected }, spreadableProps, { className: classes }),
            this.renderName(),
            this.maybeRenderContent(),
            this.props.loading ? undefined : this.props.resizeHandle));
    }
    renderName() {
        const { index, loading, name, nameRenderer, reorderHandle } = this.props;
        const dropdownMenu = this.maybeRenderDropdownMenu();
        const defaultName = React.createElement("div", { className: Classes.TABLE_TRUNCATED_TEXT }, name);
        const nameComponent = (React.createElement(LoadableContent, { loading: loading, variableLength: true }, nameRenderer == null ? defaultName : nameRenderer(name, index)));
        if (this.context.enableColumnInteractionBar) {
            return (React.createElement("div", { className: Classes.TABLE_COLUMN_NAME, title: name },
                React.createElement("div", { className: Classes.TABLE_INTERACTION_BAR },
                    reorderHandle,
                    dropdownMenu),
                React.createElement(HorizontalCellDivider, null),
                React.createElement("div", { className: Classes.TABLE_COLUMN_NAME_TEXT }, nameComponent)));
        }
        else {
            return (React.createElement("div", { className: Classes.TABLE_COLUMN_NAME, title: name },
                reorderHandle,
                dropdownMenu,
                React.createElement("div", { className: Classes.TABLE_COLUMN_NAME_TEXT }, nameComponent)));
        }
    }
    maybeRenderContent() {
        if (this.props.children === null) {
            return undefined;
        }
        return React.createElement("div", { className: Classes.TABLE_HEADER_CONTENT }, this.props.children);
    }
    maybeRenderDropdownMenu() {
        const { index, menuIcon, menuRenderer } = this.props;
        if (!CoreUtils.isFunction(menuRenderer)) {
            return undefined;
        }
        const classes = classNames(Classes.TABLE_TH_MENU_CONTAINER, {
            [Classes.TABLE_TH_MENU_OPEN]: this.state.isActive,
        });
        return (React.createElement("div", { className: classes },
            React.createElement("div", { className: Classes.TABLE_TH_MENU_CONTAINER_BACKGROUND }),
            React.createElement(Popover, { content: menuRenderer(index), position: Position.BOTTOM, className: Classes.TABLE_TH_MENU, modifiers: { preventOverflow: { boundariesElement: "window" } }, onOpened: this.handlePopoverOpened, onClosing: this.handlePopoverClosing },
                React.createElement(Icon, { icon: menuIcon }))));
    }
};
ColumnHeaderCell.defaultProps = {
    isActive: false,
    menuIcon: "chevron-down",
};
ColumnHeaderCell.contextTypes = columnInteractionBarContextTypes;
ColumnHeaderCell = tslib_1.__decorate([
    polyfill
], ColumnHeaderCell);
export { ColumnHeaderCell };
